framework/src/utils/SlepcSupport.C

```diff
@@ -356,9 +433,17 @@ moosePetscSNESFormFunction(SNES /*snes*/, Vec x, Vec r, void * ctx, Moose::Kerne
   X_global.localize(*sys.current_local_solution.get());
 
   R.zero();
-
-  eigen_problem->computeResidualType(*sys.current_local_solution.get(), R, type);
-
+  if (type == Moose::KT_EIGEN && eigen_problem->residualInitialed())
+  {
+    eigen_problem->computeResidualTypeBx(*sys.current_local_solution.get(), R, type);
+    eigen_problem->setResidualInitialed(false);
+  }
+  else
+  {
+    eigen_problem->computeResidualType(*sys.current_local_solution.get(), R, type);
+    if (type == Moose::KT_NONEIGEN)
+      eigen_problem->setResidualInitialed(true);
```

Comment:
The method is not odd at all. The` eigenvalue  solver ` is different from the normal `nonlinear` solver. We need to  evaluate the residual twice, one for the left, and the other for the right.  If the left residual is called, we are better to skip a lot of extra computations such as `multiapp transfers`, `UO` and so on. 

The idea comes from @YaqiWang